keps/sig-auth/2799-token-controller-deprecation/README.md

```diff
@@ -0,0 +1,474 @@
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required _prior to targeting to a milestone /
+release_.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a namespace-scoped and service-account-scoped binary label to
+allow users to opt in/out the provision of secret-based service account tokens
+in Token Controller. In addition, it sketches out the phases to deprecate Token
+Controller.
+
+## Motivation
+
+As BoundServiceAccountTokenVolume is GA in 1.22, pods’ service account tokens
+would be obtained via TokenRequest API and stored as projected volume. This
+change obviates the need for Token Controller’s reconciling of secret-based
+service account tokens. The secret-based tokens are [not secure by design](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/1205-bound-service-account-tokens#background)
+and the token controller is [fragile by design in some cases](https://github.com/kubernetes/kubernetes/issues/98474)
+where it is unable to handle the churns between secrets and service account
+controller loops.
+
+### Goals
+
+- Provides a knob to opt in/out the generation of secret-based token for a
+  service account
```

Comment:
Do we have a use case in mind for this?

Reply:
we want to opt out for in-tree `kube-system` SAs but not all SAs in `kube-system`.

Reply:
the current proposal was for both auto-created and manually-created secrets. admittedly, it is not friendly if we don't have an alternative "non-expiring" tokens (e.g. provides a utility for users to refresh short-lived tokens or extend TokenRequest).

if we only focus on auto-created secrets, the knobs are needed for action 1&2 because these two actions are not ok for (C). in phase 1, we only do these two actions to some kube-systen sas. or are you thinking we don't need the granularity?

Reply:
Per-namespace or per-serviceaccount controls around auto-generated secrets expose APIs I don't think we want long-term.

Clients doing C should really be doing B (which has been the documented way to manually request a service account token since Kubernetes 1.0).

Stopping auto-creating new secrets would not break use of existing secrets on upgrade, and would significantly reduce exposure on new clusters. That seems like a good place to start to me.

Once the minimum node version is 1.22 (so in 1.24), there shouldn't be pods remaining that reference secret-based tokens. Once we get to that point, we could detect use of tokens from auto-generated secrets and infer the token was extracted and used by something other than a pod. If we marked those secrets, after monitoring for "a while" (another release or so?), we could reasonably confidently delete auto-created secrets not in use by pods and not used to authenticate to the API server. Paired with advance notice that clients doing B should switch to C, and instructions for how to manually mark auto-generated secrets you scraped and didn't want to be cleaned up in a future release, I think that would eventually let us do actions 1 and 2 with minimal impact.

Reply:
sounds like a good plan. i changed the goals to reduction of auto generated secret-based service account token. action 1 is something we can do in 1.23.

one thing to mention is users can always turn a manually-created token into auto-created like token by adding the secret reference in the service account spec. also, pods can still reference secret-based tokens in 1.24 by adding them to the pod spec.